### DISCUSSION TITLE
Tips on integrating with JSON:API?

### DISCUSSION THREAD

**Discussion by madisonbullard** [TanStack/query#955] [General] [answered]
I'm investigating architecture for an upcoming project at work which might be using the [JSON:API](https://jsonapi.org/) spec, which I have never used. For reasons specific to the UI use case, I'd like to use React Query on the frontend, which I also have never used. I couldn't find any examples of the two used together in practice, and before I do too much digging I'm wondering if there are strong opinions that might help me determine whether or not to use them in conjunction.

Any shared experience would be most helpful 😄  thanks!

**Comment by tannerlinsley** [comment, marked as answer]
You can definitely use them together. We are also investigating ways to use schemas and specs like JSON APIs to make React Query even "smarter" than it already is. For now though, you can follow the documentation as is and you won't have any issues.

**Reply by hibachrach** [reply]
Has there been any development from these investigations into leveraging JSON:API's pre-existing structure?

**Reply by PCPbiscuit** [reply]
bump. Also wondering if there is anything on this topic

**Reply by osnynetto** [reply]
Me too. I'm using [jsonapi-react](https://github.com/aribouius/jsonapi-react), but it lacks many react-query features.

**Reply by rnevius** [reply]
I'm confused how this answers anything. I could be wrong, but it doesn't seem that React Query expects a JSON:API response. Instead, we have to parse the "standard" into non-standard structures which are expected by React Query...

**Reply by TkDodo** [reply]
1) the original answer is 2.5 years old

2) I don't know json-api, but whatever it is, react-query probably doesn't care. You can return a Promise from the QueryFunction. That promise can contain anything (except `undefined`), and it will be cached for you and given to you as `data` in your component.

So, idk, what else should there be integrated with? How is a response in json:api format different than regular json, or just `Promise.resolve({ some: 'object' })` ?

**Reply by rnevius** [reply]
1. Irrelevant
2. [JSON:API](https://jsonapi.org) widely used in the industry. 

The point is that it would be GREAT if patterns such as pagination were supported out of the box using this "standard".

You're right, it's all just JSON, so we can do whatever we want with it. In the case of pagination, it will mean parsing the page param from the JSON:API `link`. 

Cheers to all working on this library. It's awesome.

**Reply by TkDodo** [reply]
> The point is that it would be GREAT if patterns such as pagination were supported out of the box using this "standard".

yeah, but react-query is not a data fetching library. It's an agnostic, Promise based, async state manager. It would also be great if it supported graphql schemas and normalized caching. But all of those things I'd like to move into user-land, where you can build your own, more restricted and more powerful abstraction over react-query. It's not something we'll add to the lib itself. hope that makes sense.

**Reply by rnevius** [reply]
Makes a lot of sense, thanks @TkDodo

**Reply by rostgoat** [reply]
@rnevius what did you end up doing?

**Reply by rnevius** [reply]
@rostgoat I had to parse the response for the `nextPageParam` and then I used `useMemo` to memoize the result of flattening the pages into a single array of data (maybe not necessary). For example:

```typescript
const { data, fetchNextPage, isFetchingNextPage, isFetching } =
  useInfiniteQuery({
    queryKey: [
      'groupProjects',
      { group_id: group.id },
    ],
    queryFn: fetchProjects,
    getNextPageParam: (lastPage) =>
      lastPage?.links?.next,
    keepPreviousData: true,
    placeholderData: { pages: [], pageParams: [] },
  });

const groupProjects = useMemo(
  () => data.pages.map((page) => page.data).flat(),
  [data],
);
```
